libindy_vdr/src/pool/manager.rs

```diff
@@ -0,0 +1,212 @@
+extern crate rand;
+extern crate rmp_serde;
+
+use std::collections::HashMap;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use futures_channel::mpsc::unbounded;
+use futures_util::future::{lazy, FutureExt, LocalBoxFuture};
+use rand::seq::SliceRandom;
+
+use super::genesis::{build_node_transaction_map, build_verifiers, PoolTransactions};
+use super::networker::{
+    LocalNetworker, Networker, NetworkerEvent, NetworkerFactory, SharedNetworker,
+};
+use super::requests::{PoolRequest, PoolRequestImpl};
+use super::types::{PoolSetup, RequestHandle, Verifiers};
+
+use crate::common::error::prelude::*;
+use crate::common::merkle_tree::MerkleTree;
+use crate::config::PoolConfig;
+use crate::ledger::RequestBuilder;
+use crate::utils::base58;
+
+/// A generic verifier pool with support for creating pool transaction requests
+pub trait Pool: Clone {
+    type Request: PoolRequest;
+
+    /// Get the pool configuration for this instance
+    fn get_config(&self) -> &PoolConfig;
+
+    /// Create a new pool request instance
+    fn create_request(
+        &self,
+        req_id: String,
+        req_json: String,
+    ) -> LocalBoxFuture<'_, VdrResult<Self::Request>>;
+
+    /// Get the merkle tree representing the verifier pool transactions
+    fn get_merkle_tree(&self) -> &MerkleTree;
+
+    /// A sequence of verifier node aliases
+    fn get_node_aliases(&self) -> Vec<String>;
+
+    /// Get the size and root of the pool transactions merkle tree
+    fn get_merkle_tree_info(&self) -> (String, usize) {
+        let tree = self.get_merkle_tree();
+        (base58::encode(tree.root_hash()), tree.count())
+    }
+
+    /// Get a request builder corresponding to this verifier pool
+    fn get_request_builder(&self) -> RequestBuilder {
+        RequestBuilder::new(self.get_config().protocol_version)
+    }
+
+    /// Get the set of verifier pool transactions in JSON format
+    fn get_json_transactions(&self) -> VdrResult<Vec<String>> {
+        PoolTransactions::from(self.get_merkle_tree()).encode_json()
+    }
+
+    /// Get the summarized verifier details.
+    fn get_verifier_info(&self) -> VdrResult<Verifiers>;
+}
+
+/// The default `Pool` implementation
+#[derive(Clone)]
+pub struct PoolImpl<S: AsRef<PoolSetup> + Clone, T: Networker + Clone> {
+    setup: S,
+    networker: T,
+}
+
+/// A verifier pool instance restricted to a single thread
+pub type LocalPool = PoolImpl<Rc<PoolSetup>, LocalNetworker>;
```

Comment:
What is the reason for having a single and multi threaded pool? I assume the performance between the arc and rc is negligible here.

Reply:
I think it was mostly to allow for a single threaded web server, but I agree the performance wouldn't matter here. None of that code has changed though it's just that the file was renamed.